# Desafio Ifood

O conjunto de dados é composto por clientes da empresa Ifood com dados sobre:

- Perfis de clientes
- Preferências do produto
- Sucessos/fracassos da campanha
- Desempenho do canal

In [1]:
# Importando bibliotecas úteis para análise
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
# Caminho csv
path = 'C:/Users/João Henrique/Documents/Projetos/Desafio-Ifood/data/mkt_data.csv'

In [3]:
# Lendo csv
df = pd.read_csv(path)

In [4]:
# Mostrando o DataFrame
df

Unnamed: 0   Income  Kidhome  Teenhome  Recency  MntWines  MntFruits  \
0              0  58138.0        0         0       58       635         88   
1              1  46344.0        1         1       38        11          1   
2              2  71613.0        0         0       26       426         49   
3              3  26646.0        1         0       26        11          4   
4              4  58293.0        1         0       94       173         43   
...          ...      ...      ...       ...      ...       ...        ...   
2200        2200  61223.0        0         1       46       709         43   
2201        2201  64014.0        2         1       56       406          0   
2202        2202  56981.0        0         0       91       908         48   
2203        2203  69245.0        0         1        8       428         30   
2204        2204  52869.0        1         1       40        84          3   

      MntMeatProducts  MntFishProducts  MntSweetProducts  ...  \
0                 546              172                88  ...   
1                   6                2                 1  ...   
2                 127              111                21  ...   
3                  20               10                 3  ...   
4                 118               46                27  ...   
...               ...              ...               ...  ...   
2200              182               42               118  ...   
2201               30                0                 0  ...   
2202              217               32                12  ...   
2203              214               80                30  ...   
2204               61                2                 1  ...   

      education_Graduation  education_Master  education_PhD  MntTotal  \
0                      3.0               NaN            NaN      1529   
1                      3.0               NaN            NaN        21   
2                      3.0               NaN            NaN       734   
3                      3.0               NaN            NaN        48   
4                      NaN               NaN            5.0       407   
...                    ...               ...            ...       ...   
2200                   3.0               NaN            NaN      1094   
2201                   NaN               NaN            5.0       436   
2202                   3.0               NaN            NaN      1217   
2203                   NaN               4.0            NaN       782   
2204                   NaN               NaN            5.0       151   

      MntRegularProds  AcceptedCmpOverall  marital_status  education_level  \
0                1441                   0          Single       Graduation   
1                  15                   0          Single       Graduation   
2                 692                   0        Together       Graduation   
3                  43                   0        Together       Graduation   
4                 392                   0         Married              PhD   
...               ...                 ...             ...              ...   
2200              847                   0         Married       Graduation   
2201              428                   1        Together              PhD   
2202             1193                   1        Divorced       Graduation   
2203              721                   0        Together           Master   
2204              130                   0         Married              PhD   

      kids  expenses  
0        0      1529  
1        2        21  
2        0       734  
3        1        48  
4        1       407  
...    ...       ...  
2200     1      1094  
2201     3       436  
2202     0      1217  
2203     1       782  
2204     2       151  

[2205 rows x 44 columns]

In [5]:
# Dropando a coluna Unnamed: 0
df = df.drop("Unnamed: 0", axis=1)

In [ ]:
# Retirando os itens duplicados
df = df.drop_duplicates()

In [ ]:
# Mostrando df atualizado
df

Income  Kidhome  Teenhome  Recency  MntWines  MntFruits  \
0     58138.0        0         0       58       635         88   
1     46344.0        1         1       38        11          1   
2     71613.0        0         0       26       426         49   
3     26646.0        1         0       26        11          4   
4     58293.0        1         0       94       173         43   
...       ...      ...       ...      ...       ...        ...   
2198  26816.0        0         0       50         5          1   
2200  61223.0        0         1       46       709         43   
2202  56981.0        0         0       91       908         48   
2203  69245.0        0         1        8       428         30   
2204  52869.0        1         1       40        84          3   

      MntMeatProducts  MntFishProducts  MntSweetProducts  MntGoldProds  ...  \
0                 546              172                88            88  ...   
1                   6                2                 1             6  ...   
2                 127              111                21            42  ...   
3                  20               10                 3             5  ...   
4                 118               46                27            15  ...   
...               ...              ...               ...           ...  ...   
2198                6                3                 4             3  ...   
2200              182               42               118           247  ...   
2202              217               32                12            24  ...   
2203              214               80                30            61  ...   
2204               61                2                 1            21  ...   

      education_Graduation  education_Master  education_PhD  MntTotal  \
0                      3.0               NaN            NaN      1529   
1                      3.0               NaN            NaN        21   
2                      3.0               NaN            NaN       734   
3                      3.0               NaN            NaN        48   
4                      NaN               NaN            5.0       407   
...                    ...               ...            ...       ...   
2198                   3.0               NaN            NaN        19   
2200                   3.0               NaN            NaN      1094   
2202                   3.0               NaN            NaN      1217   
2203                   NaN               4.0            NaN       782   
2204                   NaN               NaN            5.0       151   

      MntRegularProds  AcceptedCmpOverall  marital_status  education_level  \
0                1441                   0          Single       Graduation   
1                  15                   0          Single       Graduation   
2                 692                   0        Together       Graduation   
3                  43                   0        Together       Graduation   
4                 392                   0         Married              PhD   
...               ...                 ...             ...              ...   
2198               16                   0          Single       Graduation   
2200              847                   0         Married       Graduation   
2202             1193                   1        Divorced       Graduation   
2203              721                   0        Together           Master   
2204              130                   0         Married              PhD   

      kids  expenses  
0        0      1529  
1        2        21  
2        0       734  
3        1        48  
4        1       407  
...    ...       ...  
2198     0        19  
2200     1      1094  
2202     0      1217  
2203     1       782  
2204     2       151  

[2021 rows x 43 columns]

O objetivo é fazer uma análise exploratória dos dados. Vamos primeiro responder as seguintes perguntas:

- Quantos dados temos? Linhas e colunas 
- Quais são as colunas numéricas?
- Temos duplicados na nossa base? Se tivermos, retire-os 
- Temos dados nulos nessa base? Será que eles indicam algo? O que fazer com eles? 
- Qual é a média, mediana, 25 percentil, 75 percentil, mínimo e máximo de cada uma das colunas numéricas? 

## Quantos dados temos? Linhas e colunas

In [8]:
# Usando shape para ver as linhas x colunas
df.shape

(2021, 43)

Podemos observar que temos 2021 linhas e 43 colunas

## Quais são as colunas numéricas?

In [9]:
# Usando info() para verificar as colunas numéricas
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2021 entries, 0 to 2204
Data columns (total 43 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Income                2021 non-null   float64
 1   Kidhome               2021 non-null   int64  
 2   Teenhome              2021 non-null   int64  
 3   Recency               2021 non-null   int64  
 4   MntWines              2021 non-null   int64  
 5   MntFruits             2021 non-null   int64  
 6   MntMeatProducts       2021 non-null   int64  
 7   MntFishProducts       2021 non-null   int64  
 8   MntSweetProducts      2021 non-null   int64  
 9   MntGoldProds          2021 non-null   int64  
 10  NumDealsPurchases     2021 non-null   int64  
 11  NumWebPurchases       2021 non-null   int64  
 12  NumCatalogPurchases   2021 non-null   int64  
 13  NumStorePurchases     2021 non-null   int64  
 14  NumWebVisitsMonth     2021 non-null   int64  
 15  AcceptedCmp3          2021

As colunas numéricas são aquelas que possuem o Dtype igual a int64 e float64. Podemos observar que 11 colunas são float64 e 31 colunas int64.

## Temos duplicados na nossa base? Se tivermos, retire-os

In [10]:
# Verificando se temos dados duplicados
df.duplicated().sum()

np.int64(0)

Nossos dados não possuem valores duplicados.

## Temos dados nulos nessa base? Será que eles indicam algo? O que fazer com eles?

In [11]:
# Verificando se temos dados nulos
df.isna().sum()

Income                     0
Kidhome                    0
Teenhome                   0
Recency                    0
MntWines                   0
MntFruits                  0
MntMeatProducts            0
MntFishProducts            0
MntSweetProducts           0
MntGoldProds               0
NumDealsPurchases          0
NumWebPurchases            0
NumCatalogPurchases        0
NumStorePurchases          0
NumWebVisitsMonth          0
AcceptedCmp3               0
AcceptedCmp4               0
AcceptedCmp5               0
AcceptedCmp1               0
AcceptedCmp2               0
Complain                   0
Z_CostContact              0
Z_Revenue                  0
Response                   0
Age                        0
Customer_Days              0
marital_Divorced        1807
marital_Married         1236
marital_Single          1578
marital_Together        1512
marital_Widow           1951
education_2n Cycle      1838
education_Basic         1972
education_Graduation    1006
education_Mast

Podemos ver que algumas colunas possuem valores nulos. As colunas que apresentam dados nulos são:

- marital_Divorced
- marital_Married
- marital_Single
- marital_Together
- marital_Widow
- education_2n Cycle    
- education_Basic
- education_Graduation
- education_Master
- education_PhD

Os nulos acima podem indicar que tal característica não faz parte daquela pessoa. Por exemplo, uma pessoa divorciada  
tem um determinado valor ou um valor nulo

In [12]:
# Contando os valores únicos de uma das colunas nulas. 
df["marital_Divorced"].value_counts()

marital_Divorced
1.0    214
Name: count, dtype: int64

Observando a coluna marital_Divorced, vemos que ela possuí apenas o número 1.0 que aparece 214 vezes, e o resto dos valores são nulos.  
Isso é um indício que o nulo significa que a pessoa não possuí aquela determinada variável.

Podemos substituir o NaN (valores nulos) pelos valores 0 para faltante e 1 para presente.

In [13]:
colunas_com_nulos = [
    "marital_Divorced",
    "marital_Married",
    "marital_Single",
    "marital_Together",
    "marital_Widow",
    "education_2n Cycle",
    "education_Basic",
    "education_Graduation",
    "education_Master",
    "education_PhD"
]

for coluna in colunas_com_nulos:
    df[coluna] = np.where(pd.isna(df[coluna]), 0, 1)

In [14]:
df

Income  Kidhome  Teenhome  Recency  MntWines  MntFruits  \
0     58138.0        0         0       58       635         88   
1     46344.0        1         1       38        11          1   
2     71613.0        0         0       26       426         49   
3     26646.0        1         0       26        11          4   
4     58293.0        1         0       94       173         43   
...       ...      ...       ...      ...       ...        ...   
2198  26816.0        0         0       50         5          1   
2200  61223.0        0         1       46       709         43   
2202  56981.0        0         0       91       908         48   
2203  69245.0        0         1        8       428         30   
2204  52869.0        1         1       40        84          3   

      MntMeatProducts  MntFishProducts  MntSweetProducts  MntGoldProds  ...  \
0                 546              172                88            88  ...   
1                   6                2                 1             6  ...   
2                 127              111                21            42  ...   
3                  20               10                 3             5  ...   
4                 118               46                27            15  ...   
...               ...              ...               ...           ...  ...   
2198                6                3                 4             3  ...   
2200              182               42               118           247  ...   
2202              217               32                12            24  ...   
2203              214               80                30            61  ...   
2204               61                2                 1            21  ...   

      education_Graduation  education_Master  education_PhD  MntTotal  \
0                        1                 0              0      1529   
1                        1                 0              0        21   
2                        1                 0              0       734   
3                        1                 0              0        48   
4                        0                 0              1       407   
...                    ...               ...            ...       ...   
2198                     1                 0              0        19   
2200                     1                 0              0      1094   
2202                     1                 0              0      1217   
2203                     0                 1              0       782   
2204                     0                 0              1       151   

      MntRegularProds  AcceptedCmpOverall  marital_status  education_level  \
0                1441                   0          Single       Graduation   
1                  15                   0          Single       Graduation   
2                 692                   0        Together       Graduation   
3                  43                   0        Together       Graduation   
4                 392                   0         Married              PhD   
...               ...                 ...             ...              ...   
2198               16                   0          Single       Graduation   
2200              847                   0         Married       Graduation   
2202             1193                   1        Divorced       Graduation   
2203              721                   0        Together           Master   
2204              130                   0         Married              PhD   

      kids  expenses  
0        0      1529  
1        2        21  
2        0       734  
3        1        48  
4        1       407  
...    ...       ...  
2198     0        19  
2200     1      1094  
2202     0      1217  
2203     1       782  
2204     2       151  

[2021 rows x 43 columns]

## Qual é a média, mediana, 25 percentil, 75 percentil, mínimo e máximo de cada uma das colunas numéricas?

In [15]:
# Utilizando describe para ver as medidas estatísticas
df.describe()

Income      Kidhome     Teenhome      Recency     MntWines  \
count    2021.000000  2021.000000  2021.000000  2021.000000  2021.000000   
mean    51687.258783     0.443345     0.509649    48.880752   306.492331   
std     20713.046401     0.536196     0.546393    28.950917   337.603877   
min      1730.000000     0.000000     0.000000     0.000000     0.000000   
25%     35416.000000     0.000000     0.000000    24.000000    24.000000   
50%     51412.000000     0.000000     0.000000    49.000000   178.000000   
75%     68274.000000     1.000000     1.000000    74.000000   507.000000   
max    113734.000000     2.000000     2.000000    99.000000  1493.000000   

         MntFruits  MntMeatProducts  MntFishProducts  MntSweetProducts  \
count  2021.000000      2021.000000      2021.000000       2021.000000   
mean     26.364671       166.059871        37.603662         27.268679   
std      39.776518       219.869126        54.892196         41.575454   
min       0.000000         0.000000         0.000000          0.000000   
25%       2.000000        16.000000         3.000000          1.000000   
50%       8.000000        68.000000        12.000000          8.000000   
75%      33.000000       230.000000        50.000000         34.000000   
max     199.000000      1725.000000       259.000000        262.000000   

       MntGoldProds  ...  education_2n Cycle  education_Basic  \
count   2021.000000  ...         2021.000000      2021.000000   
mean      43.921821  ...            0.090549         0.024245   
std       51.678211  ...            0.287038         0.153848   
min        0.000000  ...            0.000000         0.000000   
25%        9.000000  ...            0.000000         0.000000   
50%       25.000000  ...            0.000000         0.000000   
75%       56.000000  ...            0.000000         0.000000   
max      321.000000  ...            1.000000         1.000000   

       education_Graduation  education_Master  education_PhD     MntTotal  \
count           2021.000000       2021.000000    2021.000000  2021.000000   
mean               0.502227          0.165760       0.217219   563.789213   
std                0.500119          0.371957       0.412455   576.775749   
min                0.000000          0.000000       0.000000     4.000000   
25%                0.000000          0.000000       0.000000    55.000000   
50%                1.000000          0.000000       0.000000   343.000000   
75%                1.000000          0.000000       0.000000   964.000000   
max                1.000000          1.000000       1.000000  2491.000000   

       MntRegularProds  AcceptedCmpOverall         kids     expenses  
count      2021.000000         2021.000000  2021.000000  2021.000000  
mean        519.867392            0.302326     0.952994   563.789213  
std         554.797857            0.680812     0.746123   576.775749  
min        -283.000000            0.000000     0.000000     4.000000  
25%          42.000000            0.000000     0.000000    55.000000  
50%         288.000000            0.000000     1.000000   343.000000  
75%         883.000000            0.000000     1.000000   964.000000  
max        2458.000000            4.000000     3.000000  2491.000000  

[8 rows x 41 columns]

A média é representada pela linha mean, a mediana pelos 50%, 25 percentil por 25%, 75 percentil por 75%, mínimo pelo min e máximo pelo max.